File: python/trie/design_add_search_word.py


class WordDictionary:
    def __init__(self):
        self.children = {}
        self.is_end = False

    def addWord(self, word: str) -> None:
        node = self

        for char in word:
            if char not in node.children:
                node.children[char] = WordDictionary()

            node = node.children[char]

        node.is_end = True


    def search(self, word: str) -> bool:
        return self._search_helper(word, 0)


    def _search_helper(self, word: str, index: int) -> bool:
        if index == len(word):
            return self.is_end


        char = word[index]


        if char == '.':
            for child in self.children.values():
                if child._search_helper(word, index + 1):
                    return True
            return False

        if char in self.children:
            return self.children[char]._search_helper(word, index + 1)

        return False

File: golang/trie/design_add_search_word.go
package main

type WordDictionary struct {
	children map[rune]*WordDictionary
	isEnd    bool
}

func Constructor() WordDictionary {
	return WordDictionary{
		children: make(map[rune]*WordDictionary),
	}
}

func (wd *WordDictionary) AddWord(word string) {
	node := wd

	for _, char := range word {
		if _, found := node.children[char]; !found {
			node.children[char] = &WordDictionary{children: make(map[rune]*WordDictionary)}
		}
		node = node.children[char]
	}
	node.isEnd = true
}

func (wd *WordDictionary) Search(word string) bool {
	return wd.searchHelper([]rune(word), 0)
}

func (wd *WordDictionary) searchHelper(word []rune, index int) bool {
	if index == len(word) {
		return wd.isEnd
	}

	char := word[index]

	if char == '.' {
		for _, child := range wd.children {
			if child.searchHelper(word, index+1) {
				return true
			}
		}

		return false
	}

	if nextNode, found := wd.children[char]; found {
		return nextNode.searchHelper(word, index+1)
	}

	return false
}
